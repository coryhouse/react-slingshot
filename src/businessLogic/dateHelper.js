export default class DateHelper {
  // See tests for desired format.
  static getFormattedDateTime(date) {
    return `${date.getMonth() + 1}/${date.getDate()} ${date.getHours()}:${this.padLeadingZero(date.getMinutes())}:${this.padLeadingZero(date.getSeconds())}`;
  }

  static padLeadingZero(value) {
    return value > 9 ? value : `0${value}`;
  }
}
