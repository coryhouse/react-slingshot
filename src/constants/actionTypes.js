export const SAVE_FUEL_SAVINGS = 'SAVE_FUEL_SAVINGS';
export const CALCULATE_FUEL_SAVINGS = 'CALCULATE_FUEL_SAVINGS';
