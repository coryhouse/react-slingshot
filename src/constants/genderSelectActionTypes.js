export const GENDER_SELECTED = 'GENDER_SELECTED';