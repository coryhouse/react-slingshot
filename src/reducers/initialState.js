export default {
  fuelSavings: {
    newMpg: '',
    tradeMpg: '',
    newPpg: '',
    tradePpg: '',
    milesDriven: '',
    milesDrivenTimeframe: 'week',
    displayResults: false,
    dateModified: null,
    necessaryDataIsProvidedToCalculateSavings: false,
    savings: {
      monthly: 0,
      annual: 0,
      threeYear: 0
    }
  }
};
