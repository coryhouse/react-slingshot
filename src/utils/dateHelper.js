export function getFormattedDateTime(date = new Date()) {
  return `${date.getMonth() + 1}/${date.getDate()} ${date.getHours()}:${padLeadingZero(date.getMinutes())}:${padLeadingZero(date.getSeconds())}`;
}

export function padLeadingZero(value) {
  return value > 9 ? value : `0${value}`;
}
