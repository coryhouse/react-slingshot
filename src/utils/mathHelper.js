class MathHelper {
    static roundNumber(numberToRound, numberOfDecimalPlaces) {
        if (numberToRound === 0) {
            return 0;
        }

        if (!numberToRound) {
            return '';
        }

      const scrubbedNumber = numberToRound.toString().replace('$', '').replace(',', '');
        return Math.round(scrubbedNumber * Math.pow(10, numberOfDecimalPlaces)) / Math.pow(10, numberOfDecimalPlaces);
    }

    static addArray(values) { //adds array of values passed.
      const total = values.reduce((previousValue, currentValue) => {
        return previousValue + parseInt(this.convertToPennies(currentValue)); //do math in pennies to assure accuracy.
      }, 0);

      return total / 100; //convert back into dollars
    }

    static convertToPennies(dollarValue) {
        if (dollarValue === 0) {
            return 0;
        }

        dollarValue = parseFloat(dollarValue);
        dollarValue = this.roundNumber(dollarValue, 2); //round to 2 decimal places.
      const dollarValueContainsDecimal = (dollarValue.toString().indexOf(".") !== -1);
        return (dollarValueContainsDecimal) ? parseInt(dollarValue.toString().replace('.', '')) : parseInt(dollarValue) * 100;
    }
}

export default MathHelper;
