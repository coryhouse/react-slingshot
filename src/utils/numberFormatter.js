import MathHelper from './mathHelper';

class NumberFormatter {
	static getCurrencyFormattedNumber(value) {
        if (value === null) {
            return '';
        }

        value = this.getFormattedNumber(value);
        return '$' + value;
    }

    static getFormattedNumber(value) {
        if (value === 0) {
            return 0;
        }

        if (!value) {
            return '';
        }

        if (!this.isInt(this.scrubFormatting(value))) {
            return ''; //if it's not a number after scrubbing formatting, just return empty.
        }

      let roundedValue = MathHelper.roundNumber(value, 2); //round if more than 2 decimal points
        roundedValue = roundedValue.toString().replace(/\B(?=(\d{3})+(?!\d))/g, ","); //add commas for 1,000's. RegEx from http://stackoverflow.com/questions/2901102/how-to-print-a-number-with-commas-as-thousands-separators-in-javascript
      const roundedValueContainsDecimalPlace = (roundedValue.indexOf('.') !== -1);

        if (roundedValueContainsDecimalPlace) {
          const numbersToTheRightOfDecimal = roundedValue.split('.')[1];

            switch (numbersToTheRightOfDecimal.length) {
                case 0:
                    return roundedValue.replace('.', ''); //no decimal necessary since no numbers after decimal
                case 1:
                    return roundedValue + '0';
                default:
                    return roundedValue;
            }
        }
        return roundedValue;
    }

    static isInt(n) {
        if (n === '' || n === null) {
            return false;
        }

        return n % 1 === 0;
    }

  static scrubFormatting(value) {
        return value.toString().replace('$', '').replace(',', '').replace('.', '');
    }
}

export default NumberFormatter;
