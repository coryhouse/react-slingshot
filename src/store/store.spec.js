import * as ActionTypes from '../constants/actionTypes';

import MockDate from 'mockdate';
import { createStore } from 'redux';

import calculator from '../utils/fuelSavingsCalculator';
import {getFormattedDateTime} from '../utils/dateHelper';
import initialState from '../reducers/initialState';
import rootReducer from '../reducers';

describe('Store', () => {
  let dateModified;
  beforeAll(() => {
    // hardcoded date for consistency in tests and snapshots on all machines
    MockDate.set(new Date("1/31 23:14:01"));
    dateModified = getFormattedDateTime();
  });
  afterAll(() => MockDate.reset());

  it('should display results when necessary data is provided', () => {
    const store = createStore(rootReducer, initialState);

    const actions = [
      { type: ActionTypes.CALCULATE_FUEL_SAVINGS, dateModified, settings: store.getState(), fieldName: 'newMpg', value: 20 },
      { type: ActionTypes.CALCULATE_FUEL_SAVINGS, dateModified, settings: store.getState(), fieldName: 'tradeMpg', value: 10 },
      { type: ActionTypes.CALCULATE_FUEL_SAVINGS, dateModified, settings: store.getState(), fieldName: 'newPpg', value: 1.50 },
      { type: ActionTypes.CALCULATE_FUEL_SAVINGS, dateModified, settings: store.getState(), fieldName: 'tradePpg', value: 1.50 },
      { type: ActionTypes.CALCULATE_FUEL_SAVINGS, dateModified, settings: store.getState(), fieldName: 'milesDriven', value: 100 },
      { type: ActionTypes.CALCULATE_FUEL_SAVINGS, dateModified, settings: store.getState(), fieldName: 'milesDrivenTimeframe', value: 'month' }
    ];
    actions.forEach(action => store.dispatch(action));

    const actual = store.getState();
    const expected = {
      newMpg: 20,
      tradeMpg: 10,
      newPpg: 1.50,
      tradePpg: 1.50,
      milesDriven: 100,
      milesDrivenTimeframe: 'month',
      displayResults: false,
      dateModified,
      necessaryDataIsProvidedToCalculateSavings: true,
      savings: calculator().calculateSavings(store.getState().fuelSavings)
    };

    expect(actual.fuelSavings).toEqual(expected);
  });

  it('should not display results when necessary data is not provided', () => {
    const store = createStore(rootReducer, initialState);

    const actions = [
      { type: ActionTypes.CALCULATE_FUEL_SAVINGS, dateModified, settings: store.getState(), fieldName: 'newMpg', value: 20 },
      { type: ActionTypes.CALCULATE_FUEL_SAVINGS, dateModified, settings: store.getState(), fieldName: 'tradeMpg', value: 10 },
      // { type: ActionTypes.CALCULATE_FUEL_SAVINGS, dateModified, settings: store.getState(), fieldName: 'newPpg', value: 1.50 },
      { type: ActionTypes.CALCULATE_FUEL_SAVINGS, dateModified, settings: store.getState(), fieldName: 'tradePpg', value: 1.50 },
      { type: ActionTypes.CALCULATE_FUEL_SAVINGS, dateModified, settings: store.getState(), fieldName: 'milesDriven', value: 100 },
      { type: ActionTypes.CALCULATE_FUEL_SAVINGS, dateModified, settings: store.getState(), fieldName: 'milesDrivenTimeframe', value: 'month' }
    ];

    actions.forEach(action => store.dispatch(action));

    const actual = store.getState();

    const expected = {
      newMpg: 20,
      tradeMpg: 10,
      newPpg: '',
      tradePpg: 1.5,
      milesDriven: 100,
      milesDrivenTimeframe: 'month',
      displayResults: false,
      dateModified,
      necessaryDataIsProvidedToCalculateSavings: false,
      savings: { annual: 0, monthly: 0, threeYear: 0 }
    };


    expect(actual.fuelSavings).toEqual(expected);
  });


  it('should handle a flurry of actions', () => {
    const store = createStore(rootReducer, initialState);

    const actions = [
      { type: ActionTypes.CALCULATE_FUEL_SAVINGS, dateModified, settings: store.getState(), fieldName: 'newMpg', value: 20 },
      { type: ActionTypes.CALCULATE_FUEL_SAVINGS, dateModified, settings: store.getState(), fieldName: 'tradeMpg', value: 10 },
      { type: ActionTypes.CALCULATE_FUEL_SAVINGS, dateModified, settings: store.getState(), fieldName: 'newPpg', value: 1.50 },
      { type: ActionTypes.CALCULATE_FUEL_SAVINGS, dateModified, settings: store.getState(), fieldName: 'tradePpg', value: 1.50 },
      { type: ActionTypes.CALCULATE_FUEL_SAVINGS, dateModified, settings: store.getState(), fieldName: 'milesDriven', value: 100 },
      { type: ActionTypes.CALCULATE_FUEL_SAVINGS, dateModified, settings: store.getState(), fieldName: 'milesDrivenTimeframe', value: 'month' },
      { type: ActionTypes.CALCULATE_FUEL_SAVINGS, dateModified, settings: store.getState(), fieldName: 'newMpg', value: 20 },
      { type: ActionTypes.CALCULATE_FUEL_SAVINGS, dateModified, settings: store.getState(), fieldName: 'tradeMpg', value: 10 },
      { type: ActionTypes.CALCULATE_FUEL_SAVINGS, dateModified, settings: store.getState(), fieldName: 'newPpg', value: 1.50 },
      { type: ActionTypes.SAVE_FUEL_SAVINGS, dateModified, settings: store.getState() },
      { type: ActionTypes.CALCULATE_FUEL_SAVINGS, dateModified, settings: store.getState(), fieldName: 'tradePpg', value: 1.50 },
      { type: ActionTypes.CALCULATE_FUEL_SAVINGS, dateModified, settings: store.getState(), fieldName: 'milesDriven', value: 100 },
      { type: ActionTypes.CALCULATE_FUEL_SAVINGS, dateModified, settings: store.getState(), fieldName: 'milesDrivenTimeframe', value: 'week' },
      { type: ActionTypes.CALCULATE_FUEL_SAVINGS, dateModified, settings: store.getState(), fieldName: 'newMpg', value: 20 },
      { type: ActionTypes.CALCULATE_FUEL_SAVINGS, dateModified, settings: store.getState(), fieldName: 'tradeMpg', value: 10 },
      { type: ActionTypes.CALCULATE_FUEL_SAVINGS, dateModified, settings: store.getState(), fieldName: 'newPpg', value: 1.50 }
    ];
    actions.forEach(action => store.dispatch(action));

    calculator().calculateSavings(store.getState().fuelSavings);

    const moreActions = [
      { type: ActionTypes.CALCULATE_FUEL_SAVINGS, dateModified, settings: store.getState(), fieldName: 'tradePpg', value: 0 },
      { type: ActionTypes.CALCULATE_FUEL_SAVINGS, dateModified, settings: store.getState(), fieldName: 'milesDriven', value: 100 },
      { type: ActionTypes.CALCULATE_FUEL_SAVINGS, dateModified, settings: store.getState(), fieldName: 'milesDrivenTimeframe', value: 'year' }
    ];

    moreActions.forEach(action => store.dispatch(action));

    const actual = store.getState();
    //const expected = {
    //  newMpg: 20,
    //  tradeMpg: 10,
    //  newPpg: 1.50,
    //  tradePpg: 0,
    //  milesDriven: 100,
    //  milesDrivenTimeframe: 'year',
    //  displayResults: false,
    //  dateModified,
    //  necessaryDataIsProvidedToCalculateSavings: false,
    //  savings: lastGoodSavings
    //};
    //
    //expect(actual.fuelSavings).toEqual(expected);

    // with jest snapshots the above assertion can be replaced with this one line
    // jest will store the value in a file within ./__snapshots__
    // snapshots can/should be committed and reviewed
    // jest will also update snapshot or delete unused ones using the command `npm run test -- -u`
    expect(actual.fuelSavings).toMatchSnapshot();
  });
});
