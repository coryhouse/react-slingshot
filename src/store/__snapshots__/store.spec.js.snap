exports[`Store should handle a flurry of actions 1`] = `
Object {
  "dateModified": "1/31 23:14:01",
  "displayResults": false,
  "milesDriven": 100,
  "milesDrivenTimeframe": "year",
  "necessaryDataIsProvidedToCalculateSavings": false,
  "newMpg": 20,
  "newPpg": 1.5,
  "savings": Object {
    "annual": "$390",
    "monthly": "$32.50",
    "threeYear": "$1,170",
  },
  "tradeMpg": 10,
  "tradePpg": 0,
}
`;
